tests/unit/locking_test.cc

```diff
@@ -420,3 +422,170 @@ SEASTAR_THREAD_TEST_CASE(test_with_lock_typed_return_throwing_move_func) {
+
+        // This balances out the call to `try_lock_shared()` above.
+        sm.unlock_shared();
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_exclusive_locks) {
+    shared_mutex sm{};
+    unsigned counter = 0;
+
+    co_await coroutine::parallel_for_each(boost::irange(0, 10), coroutine::lambda([&sm, &counter] (auto&&) -> future<> {
+        const auto ulock = co_await get_unique_lock(sm);
+        BOOST_REQUIRE_EQUAL(counter, 0u);
+        ++counter;
+        co_await sleep(1ms);
+        --counter;
+        BOOST_REQUIRE_EQUAL(counter, 0u);
+    }));
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_shared_locks) {
+    shared_mutex sm{};
+    unsigned counter = 0;
+    unsigned max = 0;
+
+    co_await coroutine::parallel_for_each(boost::irange(0, 10),
+            coroutine::lambda([&sm, &counter, &max] (auto&&) -> future<> {
+        const auto slock = co_await get_shared_lock(sm);
+        ++counter;
+        max = std::max(max, counter);
+        co_await sleep(1ms);
+        --counter;
+    }));
+
+    BOOST_REQUIRE_EQUAL(counter, 0u);
+    BOOST_REQUIRE_NE(max, 0u);
+}
+
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_func_locks) {
+    shared_mutex sm;
+
+    // verify that the shared_mutex is unlocked when func fails
+    try {
+        const auto slock = co_await get_shared_lock(sm);
+        throw std::runtime_error("injected");
+    } catch (const std::runtime_error&) {
+        BOOST_REQUIRE(sm.try_lock());
+        sm.unlock();
+    } catch (...) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    try {
+        const auto ulock = co_await get_unique_lock(sm);
+        throw std::runtime_error("injected");
+    } catch (const std::runtime_error&) {
+        BOOST_REQUIRE(sm.try_lock());
+        sm.unlock();
+    } catch (...) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_lock_locks) {
+#ifdef SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+    shared_mutex sm;
+
+    // If l.lock() fails, neither the function, nor l.unlock()
+    // should be called.
+    const auto ulock = co_await get_unique_lock(sm);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await std::invoke([&] () -> future<> {
+        const auto slock = co_await get_shared_lock(sm);
+        BOOST_REQUIRE(false);
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await std::invoke([&] () -> future<> {
+        const auto ulock = co_await get_unique_lock(sm);
+        BOOST_REQUIRE(false);
+    }), std::bad_alloc);
+
+    seastar::memory::local_failure_injector().cancel();
+#endif // SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+
+    // If the macro resolves to 0, we need this to prevent compilation errors
```

Comment:
since you used ifdef and not if, I assume you mean if SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION is not defined -  not if it "resolves to 0".